src/jit/codegenxarch.cpp

```diff
@@ -1152,6 +1152,39 @@ void CodeGen::genReturn(GenTreePtr treeNode)
     // Also, there is not much to be gained by materializing it as an explicit node.
     if (compiler->compCurBB == compiler->genReturnBB)
     {
+#if defined(_TARGET_AMD64_)
+#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+        if (compiler->compMethodReturnsMultiRegRetType())
+        {
+            CORINFO_CLASS_HANDLE retClsHnd = compiler->info.compMethodInfo->args.retTypeClass;
+            ReturnTypeDesc retTypeDesc;
+            retTypeDesc.InitializeStructReturnType(compiler, retClsHnd);
+
+            unsigned regCount = retTypeDesc.GetReturnRegCount();
+
+            if (varTypeIsGC(compiler->info.compRetType))
```

Comment:
@sivarv I tried it but I have got the assertion in codegenlinear.cpp
```
System.RuntimeTypeHandle:GetMetadataImport(ref):struct: Regset after BB07 gcr=00000001 {rax}, byr=00000000 {}, regVars=00000000 {}

Assert failure(PID 27642 [0x00006bfa], Thread: 27642 [0x6bfa]): Assertion failed 'nonVarPtrRegs == RBM_NONE' in 'System.RuntimeTypeHandle:GetMetadataImport(ref):struct' (IL size 18)

    File: /home/signatov/TEST/historical_debugging/coreclr/src/jit/codegenlinear.cpp Line: 422
    Image: /home/signatov/TEST/historical_debugging/coreclr/bin/Product/Linux.x64.Debug/corerun
```

I guess that 
```
        if ((blockLastNode != nullptr) && (blockLastNode->gtOper == GT_RETURN) &&
            (varTypeIsGC(compiler->info.compRetType) ||
             (blockLastNode->gtOp.gtOp1 != nullptr && varTypeIsGC(blockLastNode->gtOp.gtOp1->TypeGet()))))
        {
            nonVarPtrRegs &= ~RBM_INTRET;
        }
```
is not working here. May be there are other reasons.